Issue #267: Inject PSR http client into BigBlueButton class, instead of using curl

donquixote opened the pull request:
See
- #267 

This probably won't work. Just a preview prototype to see the general idea.

I would like to split the class, so that the version with a http client does not have methods like `->setTimeOut()` that have no effect. We could use a shared base class for that.

donquixote committed:
Require psr/http-client and psr/http-client-factory.

donquixote committed:
Inject psr http interface objects into BigBlueButton using immutable setter.

donquixote committed:
Use a static factory instead of the immutable setter.

donquixote committed:
Make $baseUrl and $secret required in the static factory.

donquixote commented:
For cookies, see:
- #271

@GhaziTriki Let me know if we should follow up in the direction of this PR, or do things differently. Obviously this cannot be merged as-is.

GhaziTriki commented:
@donquixote looks good so far. Let's continue this pathway, and everything should be clear with an example once the implementation is finished. We need a default transport or auto-discovery when not explicitly set.

donquixote commented on src/BigBlueButton.php:
```diff
@@ -470,6 +511,10 @@ public function setJSessionId(string $jSessionId): void
     /**
+     * Sets curl options.
+     *
+     * This has no effect if the instance has a http client.
```
I think this would be "an http client"..
```suggestion
     * This has no effect if the instance has an http client.
```
and same for other similar places.

donquixote commented on src/BigBlueButton.php:
```diff
@@ -124,6 +143,28 @@ public function __construct(?string $baseUrl = null, ?string $secret = null, ?ar
+        RequestFactoryInterface $requestFactory,
+        StreamFactoryInterface $streamFactory,
+        string $baseUrl,
+        string $secret,
```
Should we have the UrlBuilder injected instead of passing base url and secret?

donquixote committed:
Require guzzle in require-dev.

donquixote committed:
Add BigBlueButtonGuzzleTest.

donquixote committed:
Remove todos in BigBlueButton::sendRequest(), which are now responsibility of the client.

donquixote committed:
Add FixturesGuzzleTest.

donquixote commented:
> We need a default transport or auto-discovery when not explicitly set.

Not sure what you mean by this. The fallback to env vars as in the constructor?
I would rather avoid this.

The main purpose of this package (I think) is to be pulled in as a dependency to be used by other code.
It is usually the application that is responsible for managing secrets. I don't know if every application wants to use the same env var mechanism that this package supports.

donquixote commented:
@GhaziTriki I think eventually we want to be able to inject the UrlBuilder.
People will need their own instance of it to call the methods.

I think it is better to discuss this separately, see
- #286.

donquixote commented on tests/BigBlueButtonGuzzleTest.php:
```diff
@@ -0,0 +1,53 @@
+ *
+ * @internal
+ */
+class BigBlueButtonGuzzleTest extends BigBlueButtonTest
```
This is a cheap way to run the same tests with curl and with a http client.

In the future we may come up with something better, perhaps.

GhaziTriki commented:
Hi @donquixote,
Tests are not passing after merges.

GhaziTriki committed:
Merge branch 'develop' into pr-273-merge

GhaziTriki committed:
Fix merge semantics and complete PSR client support

- sendRequestWithCurl keeps the array|string payload (multipart upload)
- PSR path: multipart/form-data bodies, BadResponseException on
  non-2xx, JSESSIONID capture from Set-Cookie
- FixturesGuzzleTest: load env before using it
- DocumentUrlTest: use BBB test server instead of third-party hosts

GhaziTriki committed:
Replace Guzzle with lightweight PSR-18 test client

GhaziTriki committed:
Document PSR-18 http client usage with examples
